doc/database/multi_namespace_db_instances.md

```diff
@@ -0,0 +1,495 @@
+	
+	* NPU specific services like database, swss, syncd, bgp, teamd, lldp etc which runs in a separate "namespace" created. 
+	  The method used currently to seggregate services per NPU is "linux network namespace" and there is a one-to-one 
+	  mapping between the number of NPUs and linux network namesapces created.
+
+This approach is backward compatible to the Sonic devices with single NPU. The "globalDB" service could be considered equivalent to the database docker started in the linux "host" network namespace in single NPU device. We will discuss more on the design details of database docker service in this document.
+
+The database docker started in the linux "host" namespace can be called the "global DB" service. The redis databases available here (decided by contents of database_config.json) would be APPL_DB, CONFIG_DB used to store the system wide attributes like AAA, syslog, ASIC to interface name mapping etc. 
+
+There are also database docker started per NPU in a separate "namespace". The redis databases there will include all the DB's viz. APPL_DB, CONFIG_DB, ASIC_DB, COUNTERS_DB etc and is used to store the interface/counters/state etc specific to the interfaces/features on that NPU.
+
+There will be a config_db.json file per database service, it will be named "/etc/sonic/config_db.json" for the "global DB" and /etc/sonic/config_db{NS}.json for the database service in the {NS} namespace.
+
+## Enhancements to Multi-DB design to support Multiple namespaces
+
+The file database_config.json contains the startup configuration which dictates the redis server host/port/unix_socket configurations + various available databases ( eg: APP_DB, CONFIG_DB ) etc. With the introduction of multiple namespaces, there is a need for separate database_config.json per namespace. This config file is created in the "working redis directories" during the database docker service startup. 
+
+The database docker service started in the linux network namespace {NS} will use the "/var/run/redis{NS}" as the "working redis directory" to create the various files like redis.sock, sonic-db/database_config.json etc. For "globalDB" database service it would be at "/var/run/redis".
+	
+![center](./img/redis.png)
+
+Following are the major design changes
+
+* The startup config file database_config.json is modified to have 'default' namespace DB instances + external references to the database instances present (if any) in database dockers running in other namesapces. There is a new attribute introduced, 
+
+	* "INCLUDES" to store external database_config.json file references. Each INCLUDES entry will contain the 
+	  "namespaceID" along with the "relative path" which tells the file location of database_config.json. 
+	  
+This is an example of the database_config.json with 3 external DB references in 3 namespaces refered with namespaceID "0", "1" & "2".
+
+```json
+{
```

Comment:
the json file is not clean. I suggest to have only includes in the global database json. you can put the global db configuration into a separate file and use include to include it.

Reply:
I have explicitly added details of the database_config.json file which will get generated for the global DB docker service and the NPU namespace DB docker service.  

There is no particular file like global DB database_config.json file. The current approach I have taken is to generate the json file depending on where the database docker service is running (either in linux host, or NPU namesapce )

Please find that the "INCLUDES" section is only present for the globalDB database_config.json (created at /var/run/redis/sonic-db) as the applications like snmp,pmon running in the dockers in the Linux host needs to talk to DB in other NPU namespaces. 

The database_config.json file for the NPU namespace docker won't have INCLUDES section. It will only have the default local DB instances, as the applications running in the namespace like swss, syncd, bgp etc will only talk to its local DB instances.

Reply:
Is there a use case where the applications/dockers running in a “particular NPU namespace” will need to access the DB instances in the database dockers running in either “linux host namespace” OR “other NPU namespaces” ? 

Currently my assumption is that only the applications running in the host, like snmp, CLI, pmon etc “needs to access” the DB instances running in other NPU namespaces.

Or should we make it more generic, like any application running in any docker in host/NPU namespace should be able to access the DB instance running in a different NPU namespace ? Please could you share your thoughts. 

If Yes, then I understand your ask. I would change accordinly and update the doc. Thanks

Reply:
example

```
{
   [
       {
	    "config" : "../redis/sonic-db/database_config.json"
	},
	{
	    "namespace" : "0",
	    "config" : "../redis0/sonic-db/database_config.json"
        },
	{
	    "namespace" : "1",
	    "config" : "../redis1/sonic-db/database_config.json"
        },
    ],
    "VERSION" : "1.1"
}
```

Reply:
Updated the document, by adding a new file "databases.json" which has the namespace names and the database_config.json file references. The INCLUDES will have the path for this new file.
Additional attribute added in the database_config.json file "NAMESPACE" which will have the namespace name stored.

Reply:
The document is updated, Please review.